### PR TITLE
multi-app and git submodules refactoring using Bigfoot multi-app 

### LINKED ISSUE #3235: Multi-app doc improvement
### Where on docs.platform.sh should be changed?

https://docs.platform.sh/create-apps/multi-app.html
https://docs.platform.sh/development/submodules.html

There is 2 different pages regarding submodules, one for single app and another for multi-app (and no link between them)

  

### What exactly should be updated?

# For https://docs.platform.sh/development/submodules.html

1. Clone submodules during deployment
https://docs.platform.sh/development/submodules.html#clone-submodules-during-deployment
if you just write this `.gitmodules` file with content from the doc, 
```
[submodule "app/Oro"]
	path = src/Oro
	url = https://github.com/orocrm/platform.git
[submodule "src/OroPackages/src/Oro/Bundle/EntitySerializedFieldsBundle"]
	path = src/OroPackages/src/Oro/Bundle/EntitySerializedFieldsBundle
	url = https://github.com/orocrm/OroEntitySerializedFieldsBundle.git
[submodule "src/OroB2B"]
	path = src/OroB2B
	url = https://github.com/orocommerce/orocommerce.git
```
you will get an error while deploying
```
Found 1 new commit

E: Error parsing configuration files:
    - applications: Unable to find source folder 'admin/admin-submodule' for application defined in '/admin/.platform.app.yaml', check the 'source.root: <root>' line in this file.

E: Error: Invalid configuration files, aborting build
```
as it seems that there is no action on git side to pull those submodules.
Instead of writing it manually, we need to use this git command : 
```
git submodule add --name "app/Oro" https://github.com/orocrm/platform.git "src/Oro"
git add .
git commit -m "adding app/Oro submodule"
platform push 
```

2. Removing submodules
https://docs.platform.sh/development/submodules.html#removing-submodules
```
git submodule deinit -f src/Oro
    Cleared directory 'src/Oro'
    Submodule 'app/Oro' (https://github.com/orocrm/platform.git) unregistered for path 'src/Oro'
git rm -r --cached src/Oro
git commit -m "Removed submodule."
rm -rf path_to_submodule
```

3. maybe we need to group those 2 doc pages into one, or add a link from each other to the other. 

# For https://docs.platform.sh/create-apps/multi-app.html

I try it first using submodules at the root of my sources. 

## Submodules at the root of the project

this project is using submodules at the root of sources

https://console.platform.sh/devrel-projects/a6b3w26i7qioe/multiapp-finished-submodules

with Git repo : 

- [main repo](https://github.com/platformsh-templates/bigfoot-workshop-with-api/tree/multiapp-finished-submodules)
- [Admin submodule repo](https://github.com/platformsh-templates/bigfoot-workshop-with-api-admin)
- [API submodule repo](https://github.com/platformsh-templates/bigfoot-workshop-with-api-api)
- [Gatsby submodule repo](https://github.com/platformsh-templates/bigfoot-workshop-with-api-gatsby)
- [Mercure submodule repo](https://github.com/platformsh-templates/bigfoot-workshop-with-api-mercure)

### to add them, i just did : 
```
git submodule add --name admin https://github.com/platformsh-templates/bigfoot-workshop-with-api-admin.git admin
git submodule add --name mercure https://github.com/platformsh-templates/bigfoot-workshop-with-api-mercure.git mercure
git submodule add --name api https://github.com/platformsh-templates/bigfoot-workshop-with-api-api.git api
git submodule add --name gatsby https://github.com/platformsh-templates/bigfoot-workshop-with-api-gatsby.git gatsby
git add .
git commit -m "adding submodules" && git push
platform push
```
and it generates the right `.gitmodules` file with four entries.

### Update each submodules 
```
cd admin && git pull origin main && cd ..
cd api && git pull origin main && cd ..
cd gatsby && git pull origin main && cd ..
cd mercure && git pull origin main && cd ..
git add .
git commit -m "update submodules from upstream"
platform push 
```

### Some tips

- each submodule repo needs to have its own `.platform.app.yaml` file into it

pro/con:
- as each submodule repo have their own `.platform.app.yaml`, it seems much easier for each member of the dev team to manage the way his stack needs to be deploy.

## Submodules in a subfolder of each app folder (same as described in the doc)

This environment is having subfolder in each app folder:

https://console.platform.sh/devrel-projects/a6b3w26i7qioe/multiapp-submodules-subfolders

with Git repo : 

- [main repo](https://github.com/platformsh-templates/bigfoot-workshop-with-api/tree/multiapp-submodules-subfolders)
- [Admin submodule repo](https://github.com/flovntp/bigfoot-workshop-with-api-admin)
- [API submodule repo](https://github.com/flovntp/bigfoot-workshop-with-api-api)
- [Gatsby submodule repo](https://github.com/flovntp/bigfoot-workshop-with-api-gatsby)
- [Mercure submodule repo](https://github.com/flovntp/bigfoot-workshop-with-api-mercure)

Note: Each of the forks have the same sources but without the `.platform.app.yaml` file

to add all submodules, i used : 
```
git submodule add --name admin https://github.com/flovntp/bigfoot-workshop-with-api-admin.git admin/admin-submodule 
git submodule add --name api https://github.com/flovntp/bigfoot-workshop-with-api-api.git api/api-submodule 
git submodule add --name gatsby https://github.com/flovntp/bigfoot-workshop-with-api-gatsby.git gatsby/gatsby-submodule 
git submodule add --name mercure https://github.com/flovntp/bigfoot-workshop-with-api-mercure.git mercure/mercure-submodule 
git add .
git commit -m "adding submodules" && git push
platform push
```

### Update each submodules 
```
cd admin/admin-submodule && git pull origin main && cd ../..
cd api/api-submodule && git pull origin main && cd ../..
cd gatsby/gatsby-submodule && git pull origin main && cd ../..
cd mercure/mercure-submodule && git pull origin main && cd ../..
git add .
git commit -m "update submodules from upstream"
platform push 
```

pro/con
- all PSH related files are in the same repo, so easier to manage from a devOps perspective (but harder from a dev perspective if i need to change a deploy hook by adding a new command line for example

In any case, it's not convenient to update upstream sources, maybe we need an easier way to do so but maybe it's because i'm not a git submodule expert (ping @gilzow ?)

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by flovntp**
## Why

Closes #3235

### CODE REVIEW

File: docs/src/create-apps/multi-app.md [hunk 1]
```diff
@@ -101,47 +107,65 @@ In this setup, your app configuration has to be in the top repository, not in a
...
-[submodule "app1"]
-    path = app1/app1-submodule
-    url = https://github.com/your-org/app1
-[submodule "app2"]
-    path = app2/app2-submodule
-    url = https://github.com/your-org/app2
+[submodule "admin-submodule"]
+	path = admin/admin-submodule
+	url = https://github.com/platformsh-templates/bigfoot-multiapp-admin.git
+	branch = without-platform-app-yaml
```
Comment (gilzow) [inline docs/src/create-apps/multi-app.md:140]: I would suggest the branch names for the submodules be simpler. If we aren't going to point to a typical default branch name (e.g. `main`, `master`), then we should explain why, or choose a name that doesn't give the impression the branch should _not_ contain a platform config file.

**Comment by gilzow** [comment]
My concern is that this multiapp example might be _too_ complex. At a minimum, i think there needs to be an introductory section to the example explaining the different parts, what they do, and how they relate to one another. 

As an example, I would assume that the gatsby app is the app responsible for serving the "site" to the public, but from the routes, the default site is being served by the admin app?

File: docs/src/create-apps/multi-app.md [hunk 2]
```diff
@@ -48,23 +48,29 @@ The directory with the `.platform.app.yaml` file acts as the root directory for
 For example, if you have a Drupal backend with a React frontend, you could organize the repository like this:
```
[the remark this answers is not in the document]
Comment (flovntp) [reply docs/src/create-apps/multi-app.md]: you're right, that was part of my "read it again" ;) 
sentence changed 👍

File: docs/src/create-apps/multi-app.md (hunk 1 above)
Comment (flovntp) [reply docs/src/create-apps/multi-app.md:140]: In my Bigfoot's submodule repos, i name branches regarding the difference between them (`with-platform-app-yaml` and `without-platform-app-yaml`) 
And as in this structure example, the `.platform.app.yaml` file is in the main repo, i name the submodule branch like this reflecting the fact there is no `.platform.app.yaml` file inside the submodule repo (in this case)
But you're right, there is no example in the multi-app page where we show that we could define the submodule, containing the `.platform.app.yaml` file : 
in [this branch](https://github.com/platformsh-templates/bigfoot-multiapp/tree/main), the submodule are added at the root of the main project and contain `.platform.app.yaml` file (submodule branch `with-platform-app-yaml`).

**Comment by flovntp** [comment]
@gilzow do you want me to change the project routing for having 

- `/` --> bigfoot frontend
- `/api` --> bigfoot api
- `/admin` --> API Platform 
- `/site` --> Gatsby frontend
- `mercure.{default}` --> Mercure Rocks server

**Comment by flovntp** [comment]
@AnouckColson 
the only multi-app related warning is about removing `completely` from an existing sentence.
```
## Split code source in multiple repositories (Git submodules)
You can also keep your app completely separate from the top application.
```
I don't know what to do with it :D 

```
{"message": "[Platform.editorializing] Considering avoiding 'completely' to keep the text more neutral.", "location": {"path": "docs/src/create-apps/multi-app.md", "range": {"start": {"line": 95, "column": 42}}}, "severity": "WARNING"}
```
Let me know ;)

**Comment by chadwcarlson** [comment]
@gilzow can you confirm - I think we're stuck in this loop (registering failures that exist in `main` but not in the current pull request) because we're using [`pull_request_target` and not `pull-request` as the triggering event? ](https://github.com/platformsh/platformsh-docs/blob/multi-app-submodules/.github/workflows/linting.yaml#L3)

https://stackoverflow.com/questions/74957218/what-is-the-difference-between-pull-request-and-pull-request-target-event-in-git

**Comment by AnouckColson** [comment]
Now transferred to #3302.

**Pull request closed without merging**